components/web_discovery/browser/patterns.cc

```diff
@@ -0,0 +1,329 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/web_discovery/browser/patterns.h"
+
+#include <utility>
+
+#include "base/containers/contains.h"
+#include "base/containers/fixed_flat_map.h"
+#include "base/json/json_reader.h"
+#include "base/logging.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/threading/thread_restrictions.h"
+#include "third_party/re2/src/re2/re2.h"
+
+namespace web_discovery {
+
+namespace {
+
+constexpr char kNormalPatternsKey[] = "normal";
```

Comment:
This PR has a lot of parsing code. If it's a Brave format, could we use protobuf or flatbuffer on the backend to make parsing code easy and strictly validated?

Are we going to pass some structures to renderers? If yes, maybe it makes sense to parse to mojom structure to void an extra conversion.

Reply:
@atuchin-m good idea, but it might not be straightforward since the backend needs to remain backwards compatible with the existing extension. i suppose we could maintain two formats, but that could increase the risk of errors/inconsistencies between the two. we do pass some structs to the renderer, but they are fairly slim, and are generated from the parsed patterns in memory.

Reply:
We're thinking of updating to a new DSL soon, so it should be possible to prepare protobuf or flatbuffer files when updating the CDN assets (but we will need to implement that in the extension as well while we transition to fully native). Cc @yshym